services/checks/checks.go

```diff
@@ -131,13 +130,301 @@ func (s *Service) Run(ctx context.Context) {
 
 // Checks returns available checks.
 func (s *Service) Checks() []check.Check {
-	s.m.RLock()
-	defer s.m.RUnlock()
+	s.cm.Lock()
+	defer s.cm.Unlock()
 
 	r := make([]check.Check, 0, len(s.checks))
 	return append(r, s.checks...)
 }
 
+func (s *Service) processTasks(ctx context.Context, tasks []task) {
+	ticker := time.NewTicker(time.Minute)
+
+	for {
+		if len(tasks) == 0 {
+			break
+		}
+
+		var retry []task
+		select {
+		case <-ticker.C:
+			// continue with next loop iteration
+		case <-ctx.Done():
+			return
+		}
+
+		for _, t := range tasks {
+			res, err := models.FindActionResultByID(s.db.Querier, t.resultID)
+			if err != nil {
+				s.l.Errorf("Can't find action result: %s.", err)
+				continue
+			}
+
+			if !res.Done {
+				retry = append(retry, t)
+				continue
+			}
+
+			if res.Error != "" {
+				s.l.Errorf("Action %s failed: %s.", t.resultID, res.Error)
+				continue
+			}
+
+			_, err = agentpb.UnmarshalActionQueryResult([]byte(res.Output))
+			if err != nil {
+				s.l.Errorf("Failed to parse action result with id: %s, reason: %s.", t.resultID, err)
+				continue
+			}
+
+			// TODO Execute script against returned data
+			// fmt.Println(rr)
+		}
+
+		tasks = retry
+	}
+}
+
+func (s *Service) executeChecks() {
+	mySQLChecks, postgreSQLChecks, mongoDBChecks := s.groupChecksByDB(s.checks)
+
+	var tasks []task
+	mySQLTasks, err := s.executeMySQLChecks(mySQLChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute MySQL checks: %s.", err)
+	}
+	tasks = append(tasks, mySQLTasks...)
+
+	postgreSQLTasks, err := s.executePostgreSQLChecks(postgreSQLChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute PostgreSQL checks: %s.", err)
+	}
+	tasks = append(tasks, postgreSQLTasks...)
+
+	mongoDBTasks, err := s.executeMongoChecks(mongoDBChecks)
+	if err != nil {
+		s.l.Errorf("Failed to execute MongoDB checks: %s.", err)
+	}
+	tasks = append(tasks, mongoDBTasks...)
+
+	ctx, cancel := context.WithTimeout(context.Background(), resultsTimeout)
+	defer cancel()
+
+	s.processTasks(ctx, tasks)
+}
+
+func (s *Service) executeMySQLChecks(checks []check.Check) ([]task, error) {
+	var res []task
+
+	agents, services, err := s.findAgentsAndServices(models.MySQLdExporterType)
```

Comment:
`models.FindPMMAgentsForService`, then pick the first one. We should not use exporters there.

Reply:
Oh, thanks, seems I finaly get a proper way to find agents.